### DISCUSSION TITLE
vdev replication

### DISCUSSION THREAD

**Discussion by TorbenMartin** [openzfs/zfs#16420] [General]
hello.

i have a pool:

        NAME                        STATE     READ WRITE CKSUM
        storage                     ONLINE       0     0     0
          mirror-0                  ONLINE       0     0     0
            pci-0000:2c:00.0-ata-1  ONLINE       0     0     0
            pci-0000:2c:00.0-ata-2  ONLINE       0     0     0

mirror-0 is full 70%

when i add more vdevs like:

        NAME                        STATE     READ WRITE CKSUM
        storage                     ONLINE       0     0     0
          mirror-0                  ONLINE       0     0     0
            pci-0000:2c:00.0-ata-1  ONLINE       0     0     0
            pci-0000:2c:00.0-ata-2  ONLINE       0     0     0
          mirror-1                  ONLINE       0     0     0
            pci-0000:3c:00.0-ata-1  ONLINE       0     0     0
            pci-0000:3c:00.0-ata-2  ONLINE       0     0     0
          mirror-2                  ONLINE       0     0     0
            pci-0000:4c:00.0-ata-1  ONLINE       0     0     0
            pci-0000:4c:00.0-ata-2  ONLINE       0     0     0

will zfs automatic rebalance the data between the vdevs?

70%/3 = 23,3

mirror-0 = 23,3%
mirror-1 = 23,3%
mirror-2 = 23,3%

?

and if not... ho can i rebalance the data?

**Comment by amotin** [comment]
> will zfs automatic rebalance the data between the vdevs?

No.  ZFS does not do it.  Newly written data should gradually fix imbalance with time.

> and if not... ho can i rebalance the data?

You'd need to read and rewrite the data and make sure that old data are not held by snapshots.  Also the trick won't work if dedup or block cloning is enabled unless data are complete unloaded and deleted, since their goal is to avoid extra copies.

**Reply by TorbenMartin** [reply]
> > will zfs automatic rebalance the data between the vdevs?
> 
> No. ZFS does not do it. Newly written data should gradually fix imbalance with time.
> 
> > and if not... ho can i rebalance the data?
> 
> You'd need to read and rewrite the data and make sure that old data are not held by snapshots. Also the trick won't work if dedup or block cloning is enabled unless data are complete unloaded and deleted, since their goal is to avoid extra copies.

Thanks.

1) **Newly written data should gradually fix imbalance with time.**

what will happen when i do this:


mirror-0 = 70%
mirror-1 = 100%
mirror-2 = 100%

when i write more data on it... it will be stored on mirror-0 ?

2)  **read and rewrite the data**

That means:

mirror-0 = 70% - File A
mirror-1 = 0%
mirror-2 = 0%

I just have to copy File A as File A.1 , delete File A and have something like this?

mirror-0 = 68% - File A.1/3
mirror-1 = 1%  - File A.1/3
mirror-2 = 1% - File A.1/3

**Reply by amotin** [reply]
Something like that, though it is not trivial to predict now.  I have rethinking of it on my radar, but haven't got to it yet.

**Comment by IvanVolosyuk** [comment]
If there is no dedup, you can just send / receive the data locally, if it
changes you can do that with a snapshot, then with a few incremental send /
receives, in the end you will have a brief downtime when you have to stop
the writes and move old dataset away and new in place of the old one. The
smaller the granularity of datasets the better with files being a smaller
unit of work second only to individual blocks, but it works only if you
don't have snapshots.
